binary-oracle-pair/program/src/processor.rs

```diff
@@ -0,0 +1,523 @@
+        let token_program_id_info = next_account_info(account_info_iter)?;
+        let clock_info = next_account_info(account_info_iter)?;
+        let clock = &Clock::from_account_info(clock_info)?;
+
+        if amount == 0 {
+            return Err(PoolError::InvalidAmount.into());
+        }
+
+        let user_pass_token_account = Account::unpack(&token_pass_user_account_info.data.borrow())?;
+        let user_fail_token_account = Account::unpack(&token_fail_user_account_info.data.borrow())?;
+
+        let pool = Pool::unpack(&pool_account_info.data.borrow())?;
+
+        let authority_pub_key =
+            Self::authority_id(program_id, pool_account_info.key, pool.bump_seed)?;
+        if *authority_account_info.key != authority_pub_key {
+            return Err(PoolError::InvalidAuthorityAccount.into());
+        }
+
+        if let Some(decision) = pool.decision {
+            if decision {
+                // Burn PASS tokens
+                Self::burn(
+                    &pool_account_info.key,
+                    pool.bump_seed,
+                    token_program_id_info.clone(),
+                    token_pass_mint_info.clone(),
+                    token_pass_user_account_info.clone(),
+                    authority_account_info.clone(),
+                    amount,
+                )?;
+
+                // Transfer deposit tokens from pool deposit account to user destination account
+                Self::transfer(
+                    &pool_account_info.key,
+                    pool.bump_seed,
+                    user_token_destination_account_info.clone(),
+                    authority_account_info.clone(),
+                    token_program_id_info.clone(),
+                    pool_deposit_token_account_info.clone(),
+                    amount,
+                )?;
+            } else {
+                // Burn FAIL tokens
+                Self::burn(
+                    &pool_account_info.key,
+                    pool.bump_seed,
+                    token_program_id_info.clone(),
+                    token_fail_mint_info.clone(),
+                    token_fail_user_account_info.clone(),
+                    authority_account_info.clone(),
+                    amount,
+                )?;
+
+                // Transfer deposit tokens from pool deposit account to user destination account
+                Self::transfer(
+                    &pool_account_info.key,
+                    pool.bump_seed,
+                    user_token_destination_account_info.clone(),
+                    authority_account_info.clone(),
+                    token_program_id_info.clone(),
+                    pool_deposit_token_account_info.clone(),
+                    amount,
+                )?;
+            }
+        } else {
+            let current_slot = clock.slot;
+            if current_slot < pool.mint_end_slot || current_slot > pool.decide_end_slot {
+                let possible_withdraw_amount = cmp::min(
```

Comment:
This should probably be `amount.min(user_pass_token_account.amount).min(user_fail_token_account.amount)`, otherwise we always take all of their tokens, which will probably fail if someone only approves `amount` to be taken, and `amount < possible_withdraw_amount`